src/java/org/apache/cassandra/cql3/restrictions/SimpleRestriction.java

```diff
@@ -217,7 +219,11 @@ public boolean isSupportedBy(Index index)
     @Override
     public List<ClusteringElements> values(QueryOptions options)
     {
-        assert operator == Operator.EQ || operator == Operator.IN || operator == Operator.ANN;
+        assert operator == Operator.EQ ||
+               operator == Operator.IN ||
+               operator == Operator.ANN ||
+               operator == Operator.NEQ ||
+               operator == Operator.NOT_IN;
```

Comment:
Why? They are slices operator so this method should not be called for them.

Reply:
We allowed NOT IN with ALLOW FILTERING. Remove those checks and run `SimpleRestriction#testWithUnsetValues`, you will see the path

Reply:
```
java.lang.AssertionError
	at org.apache.cassandra.cql3.restrictions.SimpleRestriction.values(SimpleRestriction.java:222)
	at org.apache.cassandra.cql3.restrictions.PartitionKeyRestrictions.nonTokenRestrictionValues(PartitionKeyRestrictions.java:197)
	at org.apache.cassandra.cql3.restrictions.PartitionKeyRestrictions.values(PartitionKeyRestrictions.java:118)
	at org.apache.cassandra.cql3.restrictions.StatementRestrictions.getPartitionKeys(StatementRestrictions.java:767)
	at org.apache.cassandra.cql3.statements.SelectStatement.getSliceCommands(SelectStatement.java:714)
	at org.apache.cassandra.cql3.statements.SelectStatement.getQuery(SelectStatement.java:412)
	at org.apache.cassandra.cql3.statements.SelectStatement.getQuery(SelectStatement.java:393)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeInternal(SelectStatement.java:589)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeLocally(SelectStatement.java:574)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeLocally(SelectStatement.java:109)
	at org.apache.cassandra.cql3.QueryProcessor.executeInternal(QueryProcessor.java:478)
	at org.apache.cassandra.cql3.CQLTester.executeFormattedQuery(CQLTester.java:1654)
	at org.apache.cassandra.cql3.CQLTester.execute(CQLTester.java:1633)
	at org.apache.cassandra.cql3.CQLTester.assertInvalidThrowMessage(CQLTester.java:2273)
	at org.apache.cassandra.cql3.CQLTester.assertInvalidThrowMessage(CQLTester.java:2252)
	at org.apache.cassandra.cql3.CQLTester.assertInvalidMessage(CQLTester.java:2237)
	at org.apache.cassandra.cql3.validation.operations.SelectSingleColumnRelationTest.testWithUnsetValues(SelectSingleColumnRelationTest.java:623)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:232)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:55)


```

Reply:
Something is wrong here. NOT_IN is a slice and slice restrictions cannot be used to build the targeted partition key they should use row filtering.

Reply:
```
        @Override
        public boolean requiresFilteringOrIndexingFor(ColumnMetadata.Kind columnKind)
        {
            return columnKind != ColumnMetadata.Kind.CLUSTERING;
        }
```
That part is probably missing.

Reply:
We have for both: 
```
@Override
public boolean requiresFilteringOrIndexingFor(ColumnMetadata.Kind columnKind)
{
        return !columnKind.isPrimaryKeyKind();
}
```

Reply:
That one is wrong, it means that we should be using filtering instead of the clustering index. An it seems that we have a problem somewhere because we should not call those for PartitionKey values.

Reply:
```
public boolean isPrimaryKeyKind()
        {
            return this == PARTITION_KEY || this == CLUSTERING;
        }
```

Reply:
Sorry, I misread it. Can you change it by  `return columnKind != ColumnMetadata.Kind.CLUSTERING;` It should solve the problem.

Reply:
Indeed, you are right. I see the confusion came from IN and EQ which we use, also for a moment, I thought about some special case with UNSET VALUE :
```
@Override
public boolean requiresFilteringOrIndexingFor(ColumnMetadata.Kind columnKind)
{
        return columnKind.isPrimaryKeyKind();
} 
```